order/conditionally_complete_lattice.lean

```diff
@@ -0,0 +1,469 @@
+  let ⟨w, hw⟩ := ‹bdd_above t› in  /-hw : ∀ (y : α), y ∈ t → y ≤ w-/
+  ⟨w, by intros; apply hw _ (‹s ⊆ t› ‹_ ∈ s›)⟩
+
+lemma bdd_below_subset (_: s ⊆ t) (_: bdd_below t) : bdd_below s :=
+  let ⟨w, hw⟩ := ‹bdd_below t› in  /-hw : ∀ (y : α), y ∈ t → w ≤ y-/
+  ⟨w, by intros; apply hw _ (‹s ⊆ t› ‹_ ∈ s›)⟩
+
+/- Boundedness of intersections of sets, in different guises, deduced from the
+subsettonicity of boundedness.-/
+lemma bdd_above_Int1 (_: bdd_above s) : bdd_above (s ∩ t) :=
+  by apply bdd_above_subset _ ‹bdd_above s›; simp
+
+lemma bdd_above_Int2 (_: bdd_above t) : bdd_above (s ∩ t) :=
+  by apply bdd_above_subset _ ‹bdd_above t›; simp
+
+lemma bdd_below_Int1 (_: bdd_below s) : bdd_below (s ∩ t) :=
+  by apply bdd_below_subset _ ‹bdd_below s›; simp
+
+lemma bdd_below_Int2 (_: bdd_below t) : bdd_below (s ∩ t) :=
+  by apply bdd_below_subset _ ‹bdd_below t›; simp
+
+end preorder
+
+/--When there is a global maximum, every set is bounded above.-/
+@[simp] lemma bdd_above_top [order_top α] (s:set α): bdd_above s :=
+  ⟨⊤, by intros; apply order_top.le_top⟩
+
+/--When there is a global minimum, every set is bounded below.-/
+@[simp] lemma bdd_above_bot [order_bot α] (s:set α): bdd_below s :=
+  ⟨⊥, by intros; apply order_bot.bot_le⟩
+
+/-When there is a max (i.e., in the class semilattice_sup), then the union of
+two bounded sets is bounded, by the maximum of the bounds for the two sets.
+With this, we deduce that finite sets are bounded by induction, and that a finite
+union of bounded sets is bounded.-/
+section semilattice_sup
+variables [semilattice_sup α] {s t: set α} {a b: α}
+
+/--The union of two sets is bounded above if and only if each of the sets is.-/
+@[simp] lemma bdd_above_union: bdd_above (s ∪ t) ↔ (bdd_above s ∧ bdd_above t) :=
+  have A: bdd_above (s ∪ t) → (bdd_above s ∧ bdd_above t) :=
+  begin
+    intro,
+    have S: bdd_above s := by apply bdd_above_subset _ ‹bdd_above (s ∪ t)›; simp,
+    have T: bdd_above t := by apply bdd_above_subset _ ‹bdd_above (s ∪ t)›; simp,
+    apply and.intro S T
+  end,
+  have B: (bdd_above s ∧ bdd_above t) → bdd_above (s ∪ t) :=
+    assume H: bdd_above s ∧ bdd_above t,
+    let ⟨⟨ws, hs⟩, ⟨wt, ht⟩⟩ := H in
+      /-hs : ∀ (y : α), y ∈ s → y ≤ ws      ht : ∀ (y : α), y ∈ s → y ≤ wt-/
+    have Bs: ∀b∈s, b ≤ ws ⊔ wt,
+      by intros; apply le_trans (hs b ‹b ∈ s›) _; simp,
+    have Bt: ∀b∈t, b ≤ ws ⊔ wt,
+      by intros; apply le_trans (ht b ‹b ∈ t›) _; simp,
+    show bdd_above (s ∪ t),
+      begin
+      apply bdd_aboveI (ws ⊔ wt),
+      intros b H_1,
+      cases H_1,
+      apply Bs _ ‹b ∈ s›,
+      apply Bt _ ‹b ∈ t›,
+      end,
+  ⟨A, B⟩
+
+/--Adding a point to a set preserves its boundedness above.-/
+@[simp] lemma bdd_above_insert: bdd_above (insert a s) ↔ bdd_above s :=
+  have A: bdd_above (insert a s) → bdd_above s :=
+    by intro; apply bdd_above_subset _ ‹bdd_above (insert a s)›; simp,
+  have B: bdd_above s → bdd_above (insert a s) := by rw[insert_eq]; finish,
+  ⟨A, B⟩
```

Comment:
I saw your remark about using this style over inlining the proofs of `A` and `B`. But what about using `show` for the same effect? i.e.
```
⟨show bdd_above (insert a s) → bdd_above s, from ...,
 show bdd_above s → bdd_above (insert a s), from ...⟩
```
I would also do the intro on the first one, so it becomes
```
⟨λ (h : bdd_above (insert a s)), show bdd_above s, from ...,
 show bdd_above s → bdd_above (insert a s), from ...⟩
```
Or just skip the intro entirely with 
```
⟨bdd_above_subset (by simp),
 show bdd_above s → bdd_above (insert a s), from ...⟩
```
(I don't know where your style guide says to insert `show` there)

Reply:
I think `bdd_above_subset (by simp)` is what Sébastien does not want to see: it does not explicitly say what you want automation to prove for you. And if automation changes and no wonger works in this case then you'll have more trouble figuring out what is failing (if I understand correctly Sébastien's point).

Reply:
That's fine - I was leaving it up to him where to insert a `show` on that one. Automation changing won't break that proof, in the sense that there are no intermediate types that can change as a result of a change in automation. `bdd_above_subset` has a fixed/known input and output type, so `simp` can only be solving `s \sub insert a s` there.

Reply:
`bdd_above_subset (by simp)` is perfect for my taste (I am not sure I understand the syntax for this one, by the way), I just kept the `show` for readability.